topology/probes/k8s/networkpolicypodlinker.go

```diff
@@ -0,0 +1,323 @@
+			}
+		}
+	}
+	return
+}
+
+func (npl *networkPolicyLinker) getIngressAllow(np *v1beta1.NetworkPolicy) (pods []metav1.Object) {
+	for _, rule := range np.Spec.Ingress {
+		for _, from := range rule.From {
+			pods = append(pods, npl.getPeerPods(from, np.Namespace)...)
+		}
+	}
+	return
+}
+
+func (npl *networkPolicyLinker) getEgressAllow(np *v1beta1.NetworkPolicy) (pods []metav1.Object) {
+	for _, rule := range np.Spec.Egress {
+		for _, to := range rule.To {
+			pods = append(pods, npl.getPeerPods(to, np.Namespace)...)
+		}
+	}
+	return
+}
+
+func fmtFieldPorts(ports []v1beta1.NetworkPolicyPort) string {
+	strPorts := []string{}
+	for _, p := range ports {
+		proto := ""
+		if p.Protocol != nil && *p.Protocol != corev1.ProtocolTCP {
+			proto = "/" + string(*p.Protocol)
+		}
+		strPorts = append(strPorts, fmt.Sprintf(":%s%s", p.Port, proto))
+	}
+	return strings.Join(strPorts, ";")
+}
+
+func getFieldPorts(np *v1beta1.NetworkPolicy, ty PolicyType) string {
+	// TODO extend logic to be able to extract the correct (per Pod object)
+	// port filter, for now all we can do is extract the ports in the case
+	// that there is only a single Ingress/egress rule (in which case we
+	// *know* the port filter applies to the specific edge).
+	ports := []v1beta1.NetworkPolicyPort{}
+	switch ty {
+	case PolicyTypeIngress:
+		if len(np.Spec.Ingress) == 1 {
+			ports = np.Spec.Ingress[0].Ports
+		}
+	case PolicyTypeEgress:
+		if len(np.Spec.Egress) == 1 {
+			ports = np.Spec.Egress[0].Ports
+		}
+	}
+	return fmtFieldPorts(ports)
+}
+
+func (npl *networkPolicyLinker) newEdgeMetadata(ty PolicyType, target PolicyTarget, point PolicyPoint) graph.Metadata {
+	m := newEdgeMetadata()
+	m.SetField("RelationType", "networkpolicy")
+	m.SetField("PolicyType", string(ty))
+	m.SetField("PolicyTarget", string(target))
+	m.SetField("PolicyPoint", string(point))
+	return m
+}
+
+func (npl *networkPolicyLinker) createLinks(np *v1beta1.NetworkPolicy, npNode, filterNode *graph.Node, ty PolicyType, target PolicyTarget, point PolicyPoint, pods []metav1.Object) (edges []*graph.Edge) {
+	podNodes := objectsToNodes(npl.graph, pods)
+	metadata := npl.newEdgeMetadata(ty, target, point)
+	for _, objNode := range podNodes {
+		if filterNode == nil || filterNode.ID == objNode.ID {
+			metadata.SetFieldAndNormalize("Ports", getFieldPorts(np, ty))
+			fields := []string{string(npNode.ID), string(objNode.ID)}
+			for k, v := range metadata {
+				fields = append(fields, k, v.(string))
+			}
+			id := graph.GenID(fields...)
+			edges = append(edges, npl.graph.CreateEdge(id, npNode, objNode, metadata, time.Now(), ""))
+		}
+	}
+	return
+}
+
+func (npl *networkPolicyLinker) getLinks(np *v1beta1.NetworkPolicy, npNode, filterNode *graph.Node) (edges []*graph.Edge) {
+	createLinks := func(ty PolicyType, target PolicyTarget, pods []metav1.Object) []*graph.Edge {
+		selectedPods := filterPodByPodSelector(npl.podCache.list(), &np.Spec.PodSelector, np.Namespace)
```

Comment:
filterPodByPodSelector only selects Pods explicitly associated, but Pods can also be implicitly associated via Namespaces defined in a NetworkPolicy.

Reply:
These are "begin" links. What you are describing is done in "getPeerPods"